src/ocl_utils_mod.f90

```diff
@@ -89,42 +94,64 @@ subroutine init_device(device, version_str, context)
-
-    write (*,'(a,i2,a,i3,a)',advance='no') &
-        'Selected Device (#',idevice,', Compute Units: ',device_cu,') - '
-    print *,device_name(1:iret)
-    deallocate(device_name)
-
-    print '(a,i2,a,i2)', 'Creating OpenCL Context for ', num_devices, &
-        ' devices of platform ', iplatform
-    ctx_props(1) = CL_CONTEXT_PLATFORM
-    ctx_props(2) = platform_ids(iplatform)
-    ctx_props(3) = 0
-    context = clCreateContext(C_LOC(ctx_props), num_devices, &
-                              C_LOC(device_ids), C_NULL_FUNPTR, C_NULL_PTR, &
-                              ierr)
-    call check_status('clCreateContext', ierr)
+    if (.not.present(device_selection)) then
+        selected_device = 1
+    else
+        selected_device = device_selection
+    endif
+    ! Get device info from all devices
+    do idevice=1, num_devices
+
+       ierr=clGetDeviceInfo(device_ids(idevice), CL_DEVICE_MAX_COMPUTE_UNITS, &
+                            sizeof(device_cu), C_LOC(device_cu), iret)
+       call check_status('clGetDeviceInfo', ierr)
+       ierr=clGetDeviceInfo(device_ids(idevice), CL_DEVICE_NAME, zero_size, &
+                            C_NULL_PTR, iret)
+       call check_status('clGetDeviceInfo', ierr)
+
+       allocate(device_name(iret), stat=iallocerr)
+       if (iallocerr.ne.0) stop 'allocate'
+
+       ierr=clGetDeviceInfo(device_ids(idevice), CL_DEVICE_NAME, &
+                            sizeof(device_name), C_LOC(device_name), iret)
+       call check_status('clGetDeviceInfo', ierr)
+
+       write (*, '(a,i2,a,i3,a)', advance='no') &
+          'Device (#', idevice,', Compute Units: ',device_cu,') - '
+       write(*,*) device_name
+       deallocate(device_name)
+    enddo
+
+    if (selected_device == 0) then
+       print '(a,i2,a,i2)', 'Creating OpenCL Context for ', num_devices, &
+          ' devices of platform ', iplatform
+       device = device_ids(1)
+       ctx_props(1) = CL_CONTEXT_PLATFORM
+       ctx_props(2) = platform_ids(iplatform)
+       ctx_props(3) = 0
+       context = clCreateContext(C_LOC(ctx_props), num_devices, &
+                                 C_LOC(device_ids), C_NULL_FUNPTR, &
+                                 C_NULL_PTR, ierr)
+       call check_status('clCreateContext', ierr)
+    elseif (selected_device > 0 .and. selected_device <= num_devices) then
+       print '(a,i1,a,i2)', 'Creating OpenCL Context for device #', &
+                selected_device, ' of platform ', iplatform
+       device = device_ids(selected_device)
+       ctx_props(1) = CL_CONTEXT_PLATFORM
+       ctx_props(2) = platform_ids(iplatform)
+       ctx_props(3) = 0
+       context = clCreateContext(C_LOC(ctx_props), 1, &
+                                 C_LOC(device_ids(selected_device)), &
+                                 C_NULL_FUNPTR, C_NULL_PTR, ierr)
+       call check_status('clCreateContext', ierr)
+    else
+        stop 'Failed to initialize OpenCL context: the device_selection' // &
```

Comment:
Sorry to be a pain but please could you extend this message to report the value of `device_selection`. I think we've already printed the number of available devices?

Reply:
That's fine (and a good thing to do), I included again the number of available devices even if it already appears somewhere before. The current message looks like:
```
Error! Failed to initialize the OpenCL context:
The device_selection parameter is '2', but it must be a number between 0 and the number of the available devices (1) in the selected platform.
```

Reply:
Thanks Sergi, that's great - it always helps to be as explicit as possible.